modules/bluefish/util/blue_velvet.cpp

```diff
@@ -20,67 +20,231 @@
-
 #include "blue_velvet.h"
+#include <common/utf.h>	
+#include <core/video_format.h>
+#include <bluefish/interop/BlueVelvetCUtils.h>
 
-#include <common/utf.h>
+#if defined(__APPLE__)
+	#include <dlfcn.h>
+#endif
 
-#include <core/video_format.h>
+#if defined(_WIN32)
+	#define GET_PROCADDR_FOR_FUNC(name, module) { name = (pFunc_##name)GetProcAddress(module, #name); if(!name) { return false; } }
+#elif defined(__APPLE__)
+	#define GET_PROCADDR_FOR_FUNC(name, module) { name = (pFunc_##name)dlsym(module, #name); if(!name) { return false; } }
+#endif
 
 namespace caspar { namespace bluefish {
-	
-CBlueVelvet4* (*BlueVelvetFactory4)() = nullptr;
-void (*BlueVelvetDestroy)(CBlueVelvet4* pObj) = nullptr;
-const char*	(*BlueVelvetVersion)() = nullptr;
-BLUE_UINT32 (*encode_hanc_frame)(struct hanc_stream_info_struct * hanc_stream_ptr, void * audio_pcm_ptr,BLUE_UINT32 no_audio_ch,BLUE_UINT32 no_audio_samples,BLUE_UINT32 nTypeOfSample,BLUE_UINT32 emb_audio_flag) = nullptr;
-BLUE_UINT32 (*encode_hanc_frame_ex)(BLUE_UINT32 card_type, struct hanc_stream_info_struct * hanc_stream_ptr, void * audio_pcm_ptr, BLUE_UINT32 no_audio_ch,	BLUE_UINT32 no_audio_samples, BLUE_UINT32 nTypeOfSample, BLUE_UINT32 emb_audio_flag) = nullptr;
 
-void blue_velvet_initialize()
-{
-#ifdef _DEBUG
-	std::string module_str = "BlueVelvet64_d.dll";
-#else
-	std::string module_str = "BlueVelvet64.dll";
+	BvcWrapper::BvcWrapper()
+	{
+		bvc = nullptr;
+		hModule = nullptr;
+		InitFunctionsPointers();
+		if (bfcFactory)
+			bvc = bfcFactory();
+	}
+
+	BvcWrapper::~BvcWrapper()
+	{
+		if (bvc)
+			bfcDestroy(bvc);
+		if (hModule)
+		{
+#if defined(_WIN32)
+			FreeLibrary(hModule);
+#elif defined(__APPLE__)
+			dlclose(hModule);
 #endif
+		}
+	}
 
-	auto module = LoadLibrary(u16(module_str).c_str());
-	if(!module)
-		LoadLibrary(u16(std::string(getenv("SystemDrive")) + "\\Program Files\\Bluefish444\\Driver\\" + module_str).c_str());
-	if(!module)
-		LoadLibrary(u16(std::string(getenv("SystemDrive")) + "\\Program Files (x86)\\BlueFish444\\Driver\\" + module_str).c_str());
-	if(!module)
-		CASPAR_THROW_EXCEPTION(file_not_found() << msg_info("Could not find BlueVelvet3.dll. Required drivers are not installed."));
-	static std::shared_ptr<void> lib(module, FreeLibrary);
-	BlueVelvetFactory4 = reinterpret_cast<decltype(BlueVelvetFactory4)>(GetProcAddress(module, "BlueVelvetFactory4"));
-	BlueVelvetDestroy  = reinterpret_cast<decltype(BlueVelvetDestroy)>(GetProcAddress(module, "BlueVelvetDestroy"));
-	BlueVelvetVersion  = reinterpret_cast<decltype(BlueVelvetVersion)>(GetProcAddress(module, "BlueVelvetVersion"));
-}
+	bool BvcWrapper::InitFunctionsPointers()
+	{
+		bool bRes = false;
 
-void blue_hanc_initialize()
-{
+#if defined(_WIN32)
 #ifdef _DEBUG
-	std::string module_str = "BlueHancUtils64_d.dll";
+		hModule = LoadLibraryExA("BlueVelvetC64_d.dll", NULL, 0);
 #else
-	std::string module_str = "BlueHancUtils64.dll";
+		hModule = LoadLibraryExA("BlueVelvetC64.dll", NULL, 0);
```

Comment:
RAII should be used by changing the type of `BvcWrapper::hModule` to `std::shared_ptr<void>` using `FreeLibrary` as the custom deleter.